### ISSUE TITLE
Recursive directory management doesn't preserve permissions

### ISSUE THREAD

**Issue by andrewclegg** [saltstack/salt#2707]
OS: Centos 6.3

Example usage:

```yaml
/opt/elasticsearch:
    file.recurse:
        - source: salt://dist/elasticsearch/0.19.10
        - user: elasticsearch
        - group: elasticsearch
        - include_empty: True
        - require:
            - user: elasticsearch
```

salt://dist/elasticsearch/0.19.10 just contains an unpacked elasticsearch distribution on the master.

On the minion, everything is set to 644 (files) / 755 (dirs) meaning that scripts are no longer executable, and have to be fixed afterwards.

**Comment by JohnHazen** [comment]
I just wanted to bump this.  After some debugging, I found this.  Perhaps a note could be added to the documentation saying "Warning: This command will not preserve file permissions."  Personally, I think file.recurse is useless without preserving permissions.  Maybe the right answer is to just remove it until it's implemented correctly, so that people know to either manage files independently, or just copy over a tarball and extract it.

**Comment by basepi** [comment]
We will not be removing `file.recurse`, it's heavily used in many salt-managed infrastructures. However, we should add a warning.

**Comment by krcroft** [comment]
@JohnHanzen, I hit this today too, and share your desire for a fully working solution. 

The suggestion to post-apply permissions & modes to an explicit subset of transferred files conflicts with the 'recurse' nature of this state, given the user of 'recurse' may not actually know apriori the set of files being recursively delivered.   I'm infact in this boat, and will need to maintain of a bunch of rules/pillars to fix things up after the transfer. 

I understand that internally solving this is difficult, but the recurse state already comes with the ability to apply dir_mode and file_modes; so an immediate fix would be to exercise this functionality on a per-file and per-dir basis, but fetching mode values from the source: salt://... directory, and then applying them on the minion.   It's not as elegant as transferring permission-metdata in the IO stream, but it would be a stop-gap.

**Comment by warrenhouzz** [comment]
+1 Do you have an ETA for this feature ?

**Comment by basepi** [comment]
Unfortunately there is no ETA at this time.

**Comment by poige** [comment]
Shame.

**Comment by thatch45** [comment]
@basepi After we review the boron release timelines I would like to consider this one for boron, but we have others that I think are a little higher on the list

**Comment by terminalmage** [comment]
As I will be implementing this, I'm going to leave these details so I don't forget them:

* Need to add a ``mode`` key to the payload when returning a payload containing a file from the fileclient. For the VCS backends, this is dependent upon support within the python bindings to get the mode.
* ``cp.cache_file`` needs to be updated to check for the mode and write the file to the minion cache with that mode. This needs to be done carefully, as the mode can be updated in a git repo while the hash of the file hasn't changed, so we can't skip checking the mode if the local hash matches. 
* File states need to be updated to stat the file in the minion cache and ensure that mode is used when deploying the file on the minion.

@thatch45 please let me know if I'm on the right track, or if there is a more elegant solution.

**Comment by basepi** [comment]
Might want to make it a general "metadata" in case we want to add more later. Make it a dict and put the mode in there. Otherwise sounds good to me.

**Comment by terminalmage** [comment]
OK. We won't be able to just dump the stat_result object into the payload because it won't serialize. Again, this information is more for my benefit when I look at this later.

**Comment by guettli** [comment]
I search a `rsync -a` like solution to copy some code to the minion. This issue here is popular and has many duplicates. I see above that something was merged in Dec 2015, but the state of the issue is still "open". 

... wait ... I found this: https://docs.saltstack.com/en/develop/ref/states/all/salt.states.rsync.html

The question is now: Does file.recurse need improvement at all, since the rsync state exists?

**Comment by jarfil** [comment]
@guettli The rsync from `state.rsync` refers only to rsync operations on the minion itself, not to the deployment of files from master to minion.

While this can be useful for making backups of directories on the minion to other directories on the minion or wherever the minion has rsync access to, it doesn't solve the problem of lost permissions while copying files from master to minion. That is, unless you give the minion ssh access to your master, or install salt-minion on the master and allow ssh access to the actual minion, or some other hack like that. Even then, those hacks would still not allow to deploy files from a filesystem that rsync has no access to, like gitfs, hgfs and such.

**Comment by guettli** [comment]
@jarfil thank you for explaining that rsync does not help here.

**Comment by jmcq** [comment]
this is a major sticking point for me.  I can't believe recursive directory copy does not preserve permissions or have an otpion to do so.

**Comment by phil123456** [comment]
this is a joke right ?

**Comment by damon-atkins** [comment]
Keep in mine puppet does not use the permission/ownership of files on the puppet master file system, they need to be listed in the Statement  i.e. Same owner, group and mode per file or directory tree.   And consideration for this sort of feature needs to also be made for Windows.

**Comment by thatch45** [comment]
Right, this is a VERY complicated issue, and I would only ever support it on *nix and explicitly enabled. I have added the file metadata to the fileserver payload though. So this is much closer to being plausible, but we still need to transfer the upstream metadata into the state evaluation on a per file basis to pull this off.

**Comment by terminalmage** [comment]
Yes, this is on my roadmap to implement for the next feature release.

**Comment by uvsmtid** [comment]
Besides generic problem of filesystem metadata propagation from master to minion, there is another aspect if we look at the problem's parts:

**`permissions = ownership + mode`**

Depending on the implementation, there are opportunities to make simpler.

I've already mentioned it in a duplicate [there](https://github.com/saltstack/salt/issues/4423#issuecomment-41463798), but it belongs here in the main discussion.

### Ownership ~ "impossible" ###

There are specific Implementation difficulties which *only* apply to user/group *ownership* because master and minion have two "completely unrelated" uid/gid - user/group databases (i.e. differences in `/etc/passwd` files, or LDAP, or ...).

Let's emphasize: **it may be completely impossible to solve *ownership* problems properly**.

Salt has nothing to do with it (unless user/group DB is also managed by the same Salt master). Just think about where ownership information may potentially come from and how *inconsistent* can it be.

### Mode ~ feasible ###

Setting only *mode* is straightforward:
set r/w/x mode based on the `source` *regardless* of the `source`'s ownership.

And even this is above what required in most cases...

### Where did we see preserving permissions from the source? ###

What are other worldwide examples with this feature?
* Many SCMs (i.e. Git, Subversion, ...) support file modes specifically for cases when scripts should be readily executable after checking them out.
* Archivers like `tar` or `zip` do exactly the same. And this is why things work out of the box (out of the tar ball).
* ...

I can bet **the majority** of use cases are narrowed down to **execution mode on *scripts* only**.

Please comment below if you really want this feature for anything else than making scripts executable.

The (other) *security*-related permission are better to be treated explicitly via states.
The catch here is that read and write modes are *closely related to ownership* and, therefore, directly associated with "impossible" to implement propagation of ownership. Executable mode, on the other hand, is *functionality*-related permission (foolproof only, easily breakable).

In other words, are you sure read and write modes are sane enough to propagate from source (potentially with the different mapping of uid/gid)?

### Conclusion ###

In short, we are likely to need "executable mode from source" only.

Is Windows still a problem in such case? Maybe someone else can comment on this - I'm not sure.

### Next step ###

(Depending on implementation)
Maybe it makes sense to split this feature request?

For example:
* (simplest, most frequent): "executable mode from source"
* (doable, less frequent): "any mode from source"
* (hardest): "ownership from source"

---

DISCLAIMER:
Maybe it was obvious to everyone already.
Maybe generic solution covers all cases anyway.
Then I'm happy to be simply ignored.

**Comment by terminalmage** [comment]
@uvsmtid The scope of this is intended to be mode only, for the very reasons you've expressed.

In order to implement this, the mode needs to be transmitted to the minion along with the payload(s) containing the file. @thatch45 has implemented this part already, from what he has said, though I haven't looked at the code to see for which fileserver backends this has been done.

The other parts of this include:

* Having the minion read this mode information and use it in states to enforce the files' modes
* Detecting mode-only changes to files

For instance, when a fileserver request is made right now, the minion will pass the checksum of the file in the request, and if the checksum of the file on the fileserver matches, the master skips re-transmitting it to avoid unneceessary retransmission of files. The mode must be added to the fileserver requests, and the master's response must be able to distinguish changes to the contents of the file from changes to the mode of the file, so that a change to just the mode does not result in retransmission.

**Comment by damon-atkins** [comment]
I agree should not pull ownership from the file system just 'x' i.e. if any x in rw-rwx--- then the script is marked as executable and "mode_exec" is used if defined, maybe add mode_dir. Otherwise mode is used.  This would keep backwards compatibility?

By the way can you do 
/srv/stuff/substuf:
  file.directory:
    - user: fred
    - group: users
    - mode: 755
    - makedirs: True
    - recurse:
      - user
      - group
      - mode
and then

/srv/stuff/substuf/xxx/foo.conf:
  file.managed:
    - source:
      - salt://foo.conf?saltenv=dev
    - user: foo
    #- group: users  no need to set this as its set above.
    - mode: '0600'

And then the second time the state is run the first statement should not flip it back to 755, followed again by going 0600

**Comment by tomduijf** [comment]
+1
I appreciate the complexity of implementing this with compatibility in mind.
 Eagerly awaiting the feature

**Comment by terminalmage** [comment]
Working on this right now. I've expanded the `os.stat()` output to the other fileserver backends, and intend to work on the fileclient and file.managed/file.recurse state modifications over the next few days. With luck I'll have a pull request opened this week.

**Comment by terminalmage** [comment]
Pull request https://github.com/saltstack/salt/pull/34807 has been opened to implement this feature. It may be modified yet before it is merged, but anyone eager to test is welcome to do so.

**Comment by guettli** [comment]
@terminalmage 
> ... but anyone eager to test

I prefer automated testing. I have no clue how to do this in salt.

**Comment by terminalmage** [comment]
@guettli integration tests for this will be added within the next week or two. That wasn't what I meant, really, the comment was more aimed towards people that would like to try out the functionality right now.

**Comment by terminalmage** [comment]
Integration test added in https://github.com/saltstack/salt/pull/34807/commits/1cd542edb8c76fc29afc839b1269b55103eb5a03.

**Comment by terminalmage** [comment]
Added a second integration test in https://github.com/saltstack/salt/pull/34807/commits/f9cd83e8bb6128e9d8e34f21db5a8ac0a07d5899.

**Comment by guettli** [comment]
Thank you very much for fixing this.

**Comment by terminalmage** [comment]
No problem. This was overdue and I'm glad to have it implemented.

**Comment by photoeye** [comment]
If this is fixed,  what version do you have to be running for it to work?  I have salt 2016.3.3 (Boron).  What flag/parameter am i supposed to run with file.recurse?

**Comment by terminalmage** [comment]
This is a new feature, it wasn't a bugfix. The feature will be in the upcoming ``Carbon`` feature release, it will not be available in the 2016.3.x release cycle. Information on how to use this feature is in the docs for the development branch of Salt (see [here](https://docs.saltstack.com/en/develop/ref/states/all/salt.states.file.html#salt.states.file.managed) and look at the docs for the ``mode`` parameter).
